### ISSUE TITLE
runtime performance metrics

### ISSUE THREAD

**Issue by ronag** [nodejs/undici#693]
What do we need for introspection and metrics?

I propose the following events:

- 'pending' - on request queued
- 'running' - on request running
- 'complete'- on request complete
- 'failed' - on request failed

Refs: https://github.com/nodejs/undici/discussions/673

**Comment by szmarczak** [comment]
1. `pending` -> `queued` (in order to avoid confusing w/ `running`)
2. Does the `failed` event have an error argument?

**Comment by ronag** [comment]
> 1. `pending` -> `queued` (in order to avoid confusing w/ `running`)

Fixed:

> 2. Does the `failed` event have an error argument?

Yes. Alternatively we could combine complete/failed and have an err argument.

**Comment by mcollina** [comment]
I think we might want to provide a "stats provider" of sort that could be attached to a Agent or a Dispatcher so that we could track all those metrics automatically - and potentially be easily collectable by prometheus.

**Comment by artur-ma** [comment]
Since undici does not use node agent, I guess it means tools like Newrelic will not track network requests correctly
Will this be enough for such tools as NR?

**Comment by mcollina** [comment]
> Since undici does not use node's built-in agent, I guess it means tools like Newrelic will not track network requests correctly

Exactly. We should look into providing some hooks for APMs in the future. 

> Will this be enough for such tools as NR?

I do not think so. This issue will provide metrics about how the process operates - data that is relatively difficult to extract from Node.js own client.

NR (and all others) will need a way to inject custom headers for each request.

**Comment by ronag** [comment]
> NR (and all others) will need a way to inject custom headers for each request.

That should be reasonably simple. Just extend `Agent` and pass another handler wrapping the user's handler. Similar to how we do redirects.

**Comment by artur-ma** [comment]
It looks great, but currently, as far as I can see, none of the methods on the handler can modify the request (headers).

```
  export interface DispatchHandlers {
    /** Invoked before request is dispatched on socket. May be invoked multiple times when a request is retried when the request at the head of the pipeline fails. */
    onConnect?(abort: () => void): void;
    /** Invoked when an error has occurred. */
    onError?(err: Error): void;
    /** Invoked when request is upgraded either due to a `Upgrade` header or `CONNECT` method. */
    onUpgrade?(statusCode: number, headers: string[] | null, socket: Duplex): void;
    /** Invoked when statusCode and headers have been received. May be invoked multiple times due to 1xx informational headers. */
    onHeaders?(statusCode: number, headers: string[] | null, resume: () => void): boolean;
    /** Invoked when response payload data is received. */
    onData?(chunk: Buffer): boolean;
    /** Invoked when response payload and trailers have been received and the request has completed. */
    onComplete?(trailers: string[] | null): void;
  }
```

The only place is inside dispatch

```
dispatch(opts, handler) {
  // modify opts and add header
  return super.dispatch(opts, new NRHandler(opts, handler))
}
```

But call to dispatch does not mean the request is executed right now (maybe it is in the queue)
or executed at all

**Comment by artur-ma** [comment]
@ronag a question regarding your example

in case im NR APM, and I want to add new Agent, I have to assume that my current globalDispatcher is `Agent`
theoretically I can do 

```javascript
class NrAgent extends getGlobalDispatcher().constructor {
  constructor(appId) {
    super()
    this.appId = appId
  }
}
```

also then I dont know what original dispatcher gets in the constructor (that is returned from `getGlobalDispatcher`)

so the other way to do it, is to wrap the original dispatcher I guess


```javascript
class NrAgent extends  Agent {
  constructor(appId, originalDispatcher) {
    super()
    this.originalDispatcher = originalDispatcher
    this.appId = appId
  }

  onHeaders() { return this.originalDispatcher.onHeaders() }
  ... 
}

setGlobalDispatcher(new NrAgent("someApp", getGlobalDispatcher()))
```

Is this expected solution ?
